juju/juju
# Add model last login to domain

Juju stores the time each user last logged in to each model. It previously did this via a collection in mongo. This PR adds a table to the access service that tracks the users last logins.

The last login column is removed from the user table and the users last login is instead calculated by selecting the most recent login by that user to any model from the new table.

Currently we are double writing the last login to mongo and to DQLite but a subsequent PR will remove the need to do that by replacing all the remaining last login calls.

The tests for ListModels in the model-manager are currently commented out so this functionality is tests. I had a go at fixing the tests but realised that ListModels itself does not work as expected and fixing that is out of scope of this PR. I have coordinated with @tlm and he will land a patch for this soon.

## Checklist




## QA steps
The QA for this PR is a little difficult since the login and permissions stories are currently in flux. Specifically, when a is granted access to a model and logs in they do not see the model when they do `juju models` (this is likely because of the changes by @tlm that move ListModels to use state). There is also the issue that adding a model and accessing it does not trigger the expected login code and therefore, new models added do not have their last login updated.

The extent of the QA possible right now is:
```
$ juju bootstrap lxd test-controller
$ juju models
Controller: test-controller

Model       Cloud/Region  Type  Status     Machines  Units  Access  Last connection
controller  lxd/default   lxd   available         1      1  admin   just now
```

In the future, it should be possible to add a model and see that the last connection is just now. For the moment it says never connected.
```
$ juju add-model default
$ juju models
Controller: test-controller

Model       Cloud/Region  Type  Status     Machines  Units  Access  Last connection
controller  lxd/default   lxd   available         1      1  admin   just now
default*    lxd/default   lxd   available         0      -  admin   never connected
```
It should also be possible to see the last logins of other users, however listing the models of other users is currently not working. You should be able to do:
```
$ juju add-user test-user
$ juju change-user-password test-user
$ juju grant test-user admin default
$ juju models --user test-user
// Or instead of listing the models as the admin, you could do
$ juju logout 
$ juju login -u test-user -c test-controller
$ juju models
```

## Links



**Jira card:** JUJU-5835
